pkg/controllers/drift/controller.go

```diff
@@ -0,0 +1,131 @@
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"knative.dev/pkg/logging"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	corecontroller "github.com/aws/karpenter-core/pkg/operator/controller"
+	"github.com/aws/karpenter/pkg/apis/v1alpha1"
+	"github.com/aws/karpenter/pkg/cloudprovider"
+)
+
+var _ corecontroller.TypedController[*v1.Node] = (*Controller)(nil)
+
+type Controller struct {
+	kubeClient    client.Client
+	cloudProvider *cloudprovider.CloudProvider
+}
+
+// NewController constructs a controller instance
+func NewController(kubeClient client.Client, cloudProvider *cloudprovider.CloudProvider) corecontroller.Controller {
+	return corecontroller.Typed[*v1.Node](kubeClient, &Controller{
+		kubeClient:    kubeClient,
+		cloudProvider: cloudProvider,
+	})
+}
+
+func (c *Controller) Name() string {
+	return "drift"
+}
+
+func (c *Controller) Reconcile(ctx context.Context, node *v1.Node) (reconcile.Result, error) {
+	provisionerName, provisionerExists := node.Labels[v1alpha5.ProvisionerNameLabelKey]
+	if !provisionerExists {
+		return reconcile.Result{}, nil
+	}
+
+	// TODO: change this label to the karpenter-core drift label
+	if drifted, ok := node.Labels["drifted"]; ok && drifted == "true" {
+		return reconcile.Result{}, nil
+	}
+
+	provisioner := &v1alpha5.Provisioner{}
+	if err := c.kubeClient.Get(ctx, types.NamespacedName{Name: provisionerName}, provisioner); err != nil {
+		return reconcile.Result{}, fmt.Errorf("getting provisioner, %w", err)
+	}
+
+	if drifted, err := c.cloudProvider.IsNodeDrifted(ctx, provisioner, node); err != nil {
+		return reconcile.Result{}, fmt.Errorf("getting drift for node, %w", err)
+	} else if drifted {
+		node.Labels["drifted"] = "true"
+	}
+
+	return reconcile.Result{RequeueAfter: 30 * time.Minute}, nil
+}
+
+func (c *Controller) Builder(ctx context.Context, m manager.Manager) corecontroller.Builder {
+	return corecontroller.Adapt(controllerruntime.
+		NewControllerManagedBy(m).
+		For(&v1.Node{}).
+		Watches(
+			&source.Kind{Type: &v1alpha5.Provisioner{}},
+			handler.EnqueueRequestsFromMapFunc(func(o client.Object) (requests []reconcile.Request) {
+				provisioner := o.(*v1alpha5.Provisioner)
+				// Ensure provisioner has a defined AWSNodeTemplate
+				nodeTemplate := &v1alpha1.AWSNodeTemplate{}
+				if err := c.kubeClient.Get(ctx, types.NamespacedName{Name: provisioner.Spec.ProviderRef.Name}, nodeTemplate); err != nil {
+					logging.FromContext(ctx).Errorf("getting AWSNodeTemplates when mapping drift watch events, %s", err)
+					return requests
+				}
```

Comment:
We're also checking if the node template exists in the reconcile method, so we can just remove this in favor of the reconcile method since we bubble up the error differently there.